students/Wieslaw_Pucilowski/lesson09/mailroom5_OO.py

```diff
@@ -0,0 +1,207 @@
+        self.name = val
+        self._donations = []
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, val):
+        if val is None:
+            self._name = None
+        else:
+            self._name = tuple(val.split())
+
+    @property
+    def donations(self):
+        return self._donations
+
+    @donations.setter
+    def donations(self, val):
+        if isinstance(val, int) or isinstance(val, float):
+            self.donations.append(val)
+        else:
+            raise ValueError("Donation should be in USD!")
+
+    @property
+    def total(self):
+        return sum(self.donations)
+
+    @property
+    def average(self):
+        return sum(self.donations) / len(self.donations)
+
+    @property
+    def number(self):
+        return len(self.donations)
+
+    @property
+    def greetings(self):
+        g1 = """
+    Ex Programmers Charity
+    1999 Heartbeat Avenue
+    11111 Fresh Spring, Alaska
+
+    Dear"""
+        g2 = "".join([' {}']*len(self.name)).format(*self.name)
+        g3 = """
+
+    Thank you so much for your generous donation of ${}
+
+    It will be put to very good use.
+
+                       Sincerely,
+                          -The Team
+
+    """.format(self.total)
+        return(g1+g2+","+g3)
+
+    @property
+    def print_greetings(self):
+        print(self.greetings)
+
+    def write_letter(self):
+        try:
+            with open("_".join(['{} ']*len(self.name)).
+                      format(*self.name)+'.txt', 'w') as f:
+                f.write(self.greetings)
+        except IOErrors as e:
+            print("""
+                Cannot write a file, cought
+                {}
+            """.format(e))
+
+
+# donors are organized in dictionary
+class DonorsCollection():
+    def __init__(self, val=None):
+        self._donors = {}
+        self.donors = val
+        self.rep = ""
+        self.wrong_option = "{:<20}".format(" - Wrong option !!!")
+        self.main_prompt = """
+        {:-^30}
+
+        1 - Send a Thank You
+        2 - Create a Report
+        3 - Send letters to everyone
+        q - Quit
+    """.format(' Main Menu ')
+        self.sub_prompt = """
+        {:-^30}
+
+        1 - Add new donor, donation
+        2 - List donors
+        q - Go to Main Menu
+
+    """.format(' Add/List donors ')
+        self.main_dispatch = {
+            '1': self.sub_menu,
+            '2': self.print_report,
+            '3': self.write_letters,
+            'q': self.quit_menu, }
+        self.sub_dispatch = {
+            '1': self.add_donor,
+            '2': self.display,
+            'q': self.quit_menu, }
+
+    def menu_selection(self, prompt, dispatch_dict):
+        while True:
+            response = input(prompt)
+            try:
+                if dispatch_dict[response]() == "exit menu":
+                    break
+            except KeyError:
+                print(response, self.wrong_option)
```

Comment:
per our conversation earlier; neither input function nor "main program flow" should reside in the data class that handles donor collections.

The simplest uses case is you are now asked to put mailroom features behind a GUI  - you would now have to refactor this class. The "separation of concern" idea is separate "user interaction code" from "data code".

Hope this helps.